README.md
[![Build Status](https://travis-ci.com/DanielleMStevens/permissR.svg?branch=main)](https://travis-ci.com/DanielleMStevens/permissR) 
<img align="right" width="140" height="150" src="https://github.com/DanielleMStevens/permissR/blob/main/github_images/permissR_logo.png">
# permissR - an R Package for finding Bacterial Permissive Sites

Danielle M. Stevens<sup>1,2</sup> and Gitta Coaker<sup>1</sup>.

<sup>1</sup>Department of Plant Pathology, University of California, Davis, USA <br />
<sup>2</sup>Integrative Genetics and Genomics Graduate Program, University of California, Davis, USA <br />

---
permissR is an R script for finding permissive sites, i.e. sites which appear to contain no functionally encoded information, proteins, regulatory regions, and mobile elements.

This package works intandom with pSelAct Express, a vector modified for gene expression in the chromosome of a permissive site. Therefore, the decisions to find such sites are ranked for this purpose.

permissR tries to identify sites which are at least 1.5 kB in length (for relatively easy, specific recombination), that contain no protein encoding genes, no mobile elements which may trigger structural changes, and are ranked by complexity to make cloning a little easier (i.e. reasonable GC-content). The outputs of this pipeline include one folder of figures which inlcudes an asessment of the overall genome, figures for regions of interest, tab-delimited text file with the ranked list of sites, and finally a tab-delimited file of any sites which share some degree of homology with the predicted site (this is only procurred if there are any). This pipeline was designed for bacteria and in its current state, not recommended on eurkaryotes.

![](/github_images/permissR_figure.png)



## 1. Install Package

To install, type the following code into your R console:


```
install.packages("devtools")
devtools::install_github("DanielleMStevens/permissR")
```

## 2. Install and run ISEScan

ISEScan can be installed easily with conda. For info on how to intall conda, [see this Github repo](https://github.com/DanielleMStevens/ROS_production_review/blob/master/process_files.md). We can then run it in one of two ways: 1) on the terminal directly or 2) on the terminal through the R-console. 

On the terminal, you'll still need to use conda to install isescan. 
```shell
conda install isescan
```


### To run it on the terminal follow below:
```
# run the following command, update the number of threads based on your computer

isescan.py --seqfile /path/to/fasta_file.fasta --output fasta_file --nthread 6
```

### If you want to run ISEScan while in R (i.e. the script will input the commands for you), you'll need to run the following commands:

The on the R-console:
```R
usethis::edit_r_profile()

# edit path to conda version of python. Anaconda should be located in your home or user directory.
Sys.setenv(PATH = paste("/home/danimstevens/anaconda3/bin", Sys.getenv("PATH"), sep=":")) 
```

Restart R and check to make sure the right version/path of python is used. 

```R
system('which python')
#/home/danimstevens/anaconda3/bin/python
```

permissR will ask the user if you want to run ISEScacn on the genome selected and will run it, but this will only work if the above steps are caried out.

## 3. Run permissR on the R-console

If the user wants to run the function similar as one would on the command line, you can feed permissR the path (relative or absolute is fine) of each file such that it looks like the following:

```
permmissR::permissR("/path/to/genbank/file.gbff", "/path/to/fasta/file.fasta")
```

Or you can run the function with no arguments and a GUI interface will appear, letting the user select each file which is required to run.
```
permissR::permissR()

# The console will alert the user to select the files to analyze. 
# This will include a fasta file of the DNA contigs, a genbank (.gbff) file, 
# and the output of ISEScan.
```


NEWS.md
# permissR 0.1.0

* Added a `NEWS.md` file to track changes to the package.


R/input-files--internal.R
#-----------------------------------------------------------------------------------------------
# Coaker Lab - Plant Pathology Department UC Davis
# Author: Danielle M. Stevens
# Last Updated: 02/22/2021
# Script Purpose: Load files to run thorough find-regions.R to search for promiscuous sites
# Inputs: fasta file, genbank file, iSEScan output
# Outputs: N/A
#-----------------------------------------------------------------------------------------------
#' Using GUI to load inital files if path is not provided by user
#' 
#' 
#' 
#' @noRd
#' @keywords internal

######################################################################
# load initial files to start script
######################################################################

.input_files <- function(){

  cat("Next Select Genbank - GBK/GBFF File\n")
  genbank_file <<- file.choose()
  
  
  cat("Select Corresponding DNA Fasta File\n")
  fasta_file <<- file.choose()
  
  
  cat("-------------------------------------------------\n")
  #return(genbank_file,fasta_file)
}



R/plot-functions--internal.R
#-----------------------------------------------------------------------------------------------
# Coaker Lab - Plant Pathology Department UC Davis
# Author: Danielle M. Stevens
# Last Updated: 06/09/2021
# Script Purpose: Find promiscuous sites in bacterial genome
# Inputs: N/A - from input-file.R 
# Outputs: Figure which plots A. total regions with annotaiton, GC content, shannon entropy, and IS/PAI elements and B. inidividual
#           plots for main hits as well as tab-delimited file output of major hit regions ranked by complexity
#-----------------------------------------------------------------------------------------------
#' 
#' 
#' 
#' 
#' @noRd
#' @keywords internal

######################################################################
# functions for subsetiing and plotting data
######################################################################


# function for plotting annotations
.annotation_plot <- function(data_to_set_up_scale, data_to_plot) {
  
  plot <- ggplot() +
    
    geom_rect(data = data_to_set_up_scale, 
              aes(xmin = start, xmax = end, ymin = 0, ymax = 0.5, fill = factor(contig)), 
              stat = "identity", fill = "white") + 
    
    geom_rect(data = data_to_plot,  
              aes(xmin = start, xmax = end, ymin = 0, ymax = 0.5, fill = factor(contig)), 
              stat = "identity", fill = "grey35") +
    
    theme_classic() +
    scale_y_continuous(expand = c(0, 0)) +
    scale_x_continuous(expand = c(0, 0)) +
    theme(legend.position = "none", axis.text.y.left = element_blank(), 
          axis.ticks.y.left = element_blank(),
          axis.title.y.left = element_text(angle = 360),
          strip.background = element_blank(),
          strip.text.x = element_blank(),
          strip.placement = "outside",
          axis.text.x = element_blank()) +
    facet_grid(. ~ factor(contig, 
                          levels = unique(forcats::fct_inorder(forcats::fct_drop(contigs_file_reformate$contig)))), 
               scales = 'free_x', space = 'free_x', switch = 'x') + 
    ylab("Genes\nPresent") 
  
  
  return(plot)
}



# function for mobile element annotations
.mobile_element_plot <- function(data_to_set_up_scale, data_to_plot){
  
  if (nrow(data_to_plot) > 0){
    plot <- ggplot() +
      geom_rect(data = data_to_set_up_scale, 
                aes(xmin = start, xmax = end, ymin = 0, ymax = 0.5, fill = factor(contig)), 
                stat = "identity", fill = "white") + 
      
      geom_rect(data = data_to_plot,  
                aes(xmin = start, xmax = end, ymin = 0, ymax = 0.5, fill = factor(contig)), 
                stat = "identity", fill = "grey35") +
      
      theme_classic() +
      scale_y_continuous(expand = c(0, 0)) +
      scale_x_continuous(expand = c(0, 0)) +
      theme(legend.position = "none", axis.text.y.left = element_blank(), 
            axis.ticks.y.left = element_blank(),
            axis.title.y.left = element_text(angle = 360),
            axis.text.x = element_blank(),
            strip.background = element_blank(),
            strip.text.x = element_blank(),
            strip.placement = "outside") +
      facet_grid(~factor(contig, 
                         levels = unique(forcats::fct_inorder(forcats::fct_drop(contigs_file_reformate$contig)))), 
                 scales = 'free_x', space = 'free_x', switch = 'x') + 
      ylab("IS Elements\nPresent")
  }
  if (nrow(data_to_plot) == 0){
    plot <- ggplot() +
      geom_rect(data = data_to_set_up_scale, 
                aes(xmin = start, xmax = end, ymin = 0, ymax = 0.5, fill = factor(contig)), 
                stat = "identity", fill = "white") + 
      geom_blank() +
      
      theme_classic() +
      scale_y_continuous(expand = c(0, 0)) +
      scale_x_continuous(expand = c(0, 0)) +
      theme(legend.position = "none", axis.text.y.left = element_blank(), 
            axis.ticks.y.left = element_blank(),
            axis.title.y.left = element_text(angle = 360),
            axis.text.x = element_blank(),
            strip.background = element_blank(),
            strip.text.x = element_blank(),
            strip.placement = "outside") +
      facet_grid(~factor(contig, 
                         levels = unique(forcats::fct_inorder(forcats::fct_drop(contigs_file_reformate$contig)))), 
                 scales = 'free_x', space = 'free_x', switch = 'x') + 
      ylab("IS Elements\nPresent")
  }
  
  return(plot)
}



# function for plotting shannon entropy
.shannon_entropy_plot <- function(data_to_set_up_scale, data_to_plot) {
  plot <- ggplot() +
    
    geom_rect(data = data_to_set_up_scale, 
              aes(xmin = start, xmax = end, ymin = 0, ymax = 0.5, fill = factor(contig)), 
              stat = "identity", fill = "white") + 
    
    geom_line(data = data_to_plot,
              aes(x = start, y = shannon.entropy, fill = factor(contig)), 
              stat = "identity", size = 0.3) +
    
    facet_grid(. ~ factor(contig, 
                          levels = unique(forcats::fct_inorder(forcats::fct_drop(contigs_file_reformate$contig)))), scales = 'free_x', space = 'free_x', switch = 'x') + 
    theme_classic() +
    scale_x_continuous(expand = c(0, 0)) +
    ylim(1.4,2.0) +
    theme(axis.text.x = element_blank(), 
          axis.title.x = element_blank(),
          axis.text.y = element_text(family = "Arial", size = 11, color = "black"),
          axis.title.y.left = element_text(angle = 360),
          strip.background = element_blank(),
          strip.text.x = element_blank(),
          strip.placement = "outside") +
    ylab("Shannon\nEntropy") 
  
  return(plot)
}


#data_to_set_up_scale
# function for plotting gc content with contig info
.gc_content_plot <- function(data_to_set_up_scale, data_to_plot, sites_to_plot) {
  
  plot <- ggplot() +
    
    geom_rect(data = data_to_set_up_scale, 
              aes(xmin = start, xmax = end, ymin = 0, ymax = 1),
              fill = "white") +
    
    geom_line(data = data_to_plot, 
              aes(x = start, y =  gc.content, fill = factor(contig)), 
              stat = "identity", size = 0.3, color = "black") + 
    
    geom_rect(data = sites_to_plot,  
              aes(xmin = position.1, xmax = position.2, ymin = 0, ymax = 1, fill = factor(contig)), 
              stat = "identity", fill = "#0072B2", alpha = 0.8) +
    
    facet_grid(. ~ factor(contig, 
                          levels = unique(forcats::fct_inorder(forcats::fct_drop(data_to_plot$contig)))), scales = 'free_x', space = 'free_x', switch = 'x') + 
    theme_classic() +
    scale_y_continuous(expand = c(0, 0), limits = c(0,1)) +
    scale_x_continuous(expand = c(0, 0)) +
    theme(axis.text.x = element_text(family = "Arial", size = 11, color = "black", 
                                     angle = 90), 
          axis.text.y = element_text(family = "Arial", size = 11, color = "black"),
          axis.title.y.left = element_text(angle = 360),
          strip.background = element_rect(fill = "#B5EB5A"),
          legend.position = "none",
          strip.text.x = element_text(angle = 90, family = "Arial", 
                                      size = 14, color = "black"),
          strip.placement = "outside") +
    ylab("%GC") +
    xlab("\nPosition along Chromosome") 
  
  return(plot)
}




R/reformate-genbank--internal.R
#-----------------------------------------------------------------------------------------------
# Coaker Lab - Plant Pathology Department UC Davis
# Author: Danielle M. Stevens
# Last Updated: 02/22/2021
# Script Purpose: Load files to run thorough find-regions.R to search for promiscuous sites
# Inputs: fasta file, genbank file, iSEScan output
# Outputs: N/A
#-----------------------------------------------------------------------------------------------
#' Using GUI to load inital files if path is not provided by user
#' 
#' 
#' 
#' @noRd
#' @keywords internal

######################################################################
# load initial files to start script
######################################################################

.reformate_genbank <- function(){
  
  
  if(grepl("\\b.gbff\\b", genbank_file) == TRUE){
    Problem_annotations <- c("translation","organism","linkage_evidence")
    
    for (i in 1:length(annotation_file$FEATURES)){
      if(length(annotation_file$FEATURES[[i]]) == 0){
        next
      }
      
      if(length(annotation_file$FEATURES[[i]]) != 0){
        if(any(grepl(paste(Problem_annotations, collapse = "|"), colnames(annotation_file$FEATURES[[i]]))) == FALSE){
          if(any(grepl("locus_tag", colnames(annotation_file$FEATURES[[i]]))) == FALSE){
            next
          }
          
          # extract gene name when possible, otherwise use locus-tag
          if(any(grepl("gene",colnames(annotation_file$FEATURES[[i]]))) == TRUE){
            gene_name <- annotation_file$FEATURES[[i]]$gene
            locus_tag <- annotation_file$FEATURES[[i]]$locus_tag
          }
          if(any(grepl("gene",colnames(annotation_file$FEATURES[[i]]))) == FALSE){
            gene_name <- annotation_file$FEATURES[[i]]$locus_tag
            locus_tag <- annotation_file$FEATURES[[i]]$locus_tag
          }
          if(annotation_file$FEATURES[[i]]$type == "regulatory"){
            gene_name <- annotation_file$FEATURES[[i]]$regulatory_class
            locus_tag <- annotation_file$FEATURES[[i]]$db_xref
          }
          
          temp_df <- data.frame("gene" = gene_name, 
                                "locus-tag" = locus_tag,
                                'type' = annotation_file$FEATURES[[i]]$type,
                                'strand' = annotation_file$FEATURES[[i]]$strand,
                                'start' = annotation_file$FEATURES[[i]]$start,
                                'end' = annotation_file$FEATURES[[i]]$end)
          
          anno_file_reformate <- rbind(anno_file_reformate, temp_df)
          
        }
      }
    }
    
    # occasuinally some CDS sequences get through, this is to remove those redudancies
    anno_file_reformate <- anno_file_reformate[anno_file_reformate$type %in% c("gene","regulatory"),]
  }
  
  if(grepl("\\b.gbk\\b", genbank_file) == TRUE){
    Problem_annotations <- c("organism","linkage_evidence")
    for (i in 1:length(annotation_file$FEATURES)){
      if(length(annotation_file$FEATURES[[i]]) == 0){
        next
      }
      
      if(length(annotation_file$FEATURES[[i]]) != 0){
        if(any(grepl(paste(Problem_annotations, collapse = "|"), colnames(annotation_file$FEATURES[[i]]))) == FALSE){
          #if(any(grepl("locus_tag", colnames(annotation_file$FEATURES[[i]]))) == FALSE){
          #  next
          #}
          
          # extract gene name when possible, otherwise use locus-tag
          if(any(grepl("gene",colnames(annotation_file$FEATURES[[i]]))) == FALSE){
            gene_name <- annotation_file$FEATURES[[i]]$locus_tag
            locus_tag <- annotation_file$FEATURES[[i]]$locus_tag
          }
          if(annotation_file$FEATURES[[i]]$type == "regulatory"){
            gene_name <- annotation_file$FEATURES[[i]]$regulatory_class
            locus_tag <- annotation_file$FEATURES[[i]]$db_xref
          }
          
          temp_df <- data.frame("gene" = gene_name, 
                                "locus-tag" = locus_tag,
                                'type' = annotation_file$FEATURES[[i]]$type,
                                'strand' = annotation_file$FEATURES[[i]]$strand,
                                'start' = annotation_file$FEATURES[[i]]$start,
                                'end' = annotation_file$FEATURES[[i]]$end)
          
          anno_file_reformate <- rbind(anno_file_reformate, temp_df)
          #print(anno_file_reformate)
          
        }
      }
    }
  }
  return(anno_file_reformate)

}

R/MBE-files--internal.R
#-----------------------------------------------------------------------------------------------
# Coaker Lab - Plant Pathology Department UC Davis
# Author: Danielle M. Stevens
# Last Updated: 06/09/2021
# Script Purpose: Load files to run thorough find-regions.R to search for promiscuous sites
# Inputs: fasta file, genbank file, iSEScan output
# Outputs: N/A
#-----------------------------------------------------------------------------------------------
#' tries to run ISEScan and asks user for output to load into script
#' 
#' 
#' 
#' @noRd
#' @keywords internal

######################################################################
# library packages need to load
######################################################################

.MBE_files <- function(){
  
  library(parallel)
  
  check_IS_Scan <- readline(prompt = "Do you need to run ISEScan? (y/n): ")
  
  if (check_IS_Scan == "y"){
    number_of_cores <- parallel::detectCores()/2
    update_file_name <- gsub(dirname(fasta_file), "", fasta_file)
    update_file_name <- gsub("/","",update_file_name)
    update_file_name <- gsub(".fasta","",update_file_name)

    
    system(paste('isescan.py --seqfile', fasta_file, '--output', update_file_name, '--nthread', number_of_cores, sep = ' '))
  }
  
  
  
  
  var1_IS <- readline(prompt = "Is there an output from ISEScan? (y/n): ")
  if (var1_IS == "y"){
    var1_IS <<- TRUE
    cat("Select .fasta.out file from ISEScan Output\n")
    is_element_file <- file.choose()
  }
  if (var1_IS == "n"){
    var1_IS <<- FALSE
  }
}

cat("-------------------------------------------------\n")


#var2_GI <- readline(prompt = "Is there an output from IslandViewer4? (y/n): ")
#if (var2_GI == "y"){
#  var2_GI <- TRUE
#  print("Select IslandViewer4 file Output")
#  GI_file <- file.choose()
#}
#if (var2_GI == "n"){
#  var2_GI <- FALSE
#}

#cat("-------------------------------------------------\n")


